### PR TITLE
Database bootstrapping

### PR DESCRIPTION
**Pull request by dandds**
## Script for setting up database user, schema, and seed data.

This script is for bootstrapping the initial database. It can be run via
a container, but requires that a Postgres superuser's credentials be
provided via our normal config. That way the superuser can provision a
less-privileged user for the application's database connection.

## secrets-tool command for bootstrapping database.

This additional secrets-tool command can be used to run the database
bootsrapping script (`script/database_setup.py`) inside an ATAT docker
container against the Azure database. It sources the necessary keys from
Key Vault.

## Create database with separate script.

Creating the ATAT database requires a separate connection to one of the
default Postgres databases, like `postgres`. This updates the scripts
and secrets-tool command to handle creating the database. It also
removes database creation from Terraform and updates the documentation.

## terraform plan

Screen Shot 2020-01-25 at 12 27 03 PM

Note: to test this, you currently have to whitelist your IP with the Postgres database:

Screen Shot 2020-01-25 at 12 29 57 PM

### CODE REVIEW

File: script/database_setup.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+from atst.app import make_config, make_app
+from atst.database import db
+from atst.domain.users import Users
+from atst.models import User
+from reset_database import reset_database
+
+
+def database_setup(username, password, dbname, ccpo_users):
+    """docstring for database_setup"""
```
Comment (graham-dds) [inline script/database_setup.py]: Is this the actual docstring? Or the quasi `Lorem ipsum` equivalent?

File: notes.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+- for setting up the database:
```
Comment (graham-dds) [inline notes.md]: Did you mean to commit this file?

File: terraform/secrets-tool/commands/database.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    logger.info("starting docker process")
+
+    create_database_cmd = (
+        f"docker run -e PGHOST='{dbhost}'"
+        +f" -e PGPASSWORD='{root_password}'"
+        +f" -e PGUSER='{root_name}@{dbhost}'"
+        +f" -e PGDATABASE='{dbname}'"
+        +f" -e PGSSLMODE=require"
+        +f" {container}"
+        +f" .venv/bin/python script/create_database.py {dbname}"
```
Comment (graham-dds) [inline terraform/secrets-tool/commands/database.py]: I don't _think_ you need `+` on every line for python to [concatenate these strings](https://docs.python.org/3/reference/lexical_analysis.html#string-literal-concatenation). Same in `seed_database_cmd` below and `_create_database_user` in `database_setup.py`

File: users.yml [hunk 4]
```diff
@@ -0,0 +1,3 @@
+- dod_id: "2323232323"
```
Comment (graham-dds) [inline users.yml]: What does this file do? Is this just for the development environment for these commands to run?

File: notes.md (hunk 2 above)
Comment (dandds) [reply notes.md]: Sure didn't

File: terraform/secrets-tool/commands/database.py (hunk 3 above)
Comment (dandds) [reply terraform/secrets-tool/commands/database.py]: Yeah, I got a slightly different syntax error early on that made me think I needed it, but you're right.

File: users.yml (hunk 4 above)
Comment (dandds) [reply users.yml]: It was just a sample for the `secrets-tool` command I committed by accident.

File: script/database_setup.py (hunk 1 above)
Comment (dandds) [reply script/database_setup.py]: Yeah, lorem ipsum. Will delete.

**Comment by dandds** [comment]
@graham-dds Thanks, should be cleaned up now.

File: script/database_setup.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    meta.reflect()
+
+    trans = conn.begin()
+    engine = trans.connection.engine
+
+    engine.execute(
+        f"CREATE ROLE {username} WITH LOGIN NOSUPERUSER INHERIT NOCREATEDB NOCREATEROLE NOREPLICATION PASSWORD '{password}';\n"
+        f"GRANT ALL PRIVILEGES ON DATABASE {dbname} TO {username};\n"
+        f"ALTER DEFAULT PRIVILEGES IN SCHEMA public GRANT ALL PRIVILEGES ON TABLES TO {username}; \n"
+        f"ALTER DEFAULT PRIVILEGES IN SCHEMA public GRANT ALL PRIVILEGES ON SEQUENCES TO {username}; \n"
```
[the remark this answers is not in the document]
Comment (JRNPW) [reply script/database_setup.py]: I created [this Pivotal Tracker card](https://www.pivotaltracker.com/story/show/170983955) to address this post-launch.

[the remark this answers is not in the document]
Comment (whit1206) [reply script/database_setup.py:46]: This is now ticket AT-4977

**Pull request merged**